docs/guide/snaps-development-guide.md

```diff
@@ -1,47 +1,290 @@
+Meanwhile, the CLI will update some of the fields for you, e.g. `source.shasum` whenever you run `mm-snap build` (by default) or `mm-snap manifest --fix`.
+
+### The Snap Bundle File
+
+Because of the way snaps are executed, they must be published as a single `.js` file containing the entire source code and all dependencies.
+Moreover, the [snaps execution environment](#the-snap-execution-environment) has no DOM, no Node.js APIs, and (needless to say) no filesystem access, so anything that relies on the DOM won't work, and any Node builtins have to be bundled along with the snap as well.
+If this sounds like a lot to worry about, `mm-snap build` is your friend, because it will bundle your snap for you using [Browserify](https://browserify.org).
+
+`mm-snap build` will find all dependencies via your specified main entry point and output a bundle file to your specified output path.
 
-To get started with Snaps, you first have to download [MetaMask Flask](https://metamask.io/flask).
+## Developing a Snap
 
-While this guide is under construction, we **highly recommend** consulting the [readme of `@metamask/snaps-cli`](https://www.npmjs.com/package/@metamask/snaps-cli), the command line utility we created to help with snap development.
+Snaps exist in order to modify the functionality of MetaMask at runtime while only asking the user for permission.
+As we have seen in the [introduction to snaps](./snaps.html) and this guide, snaps can:
 
-In addition, you can review the growing number of [example snaps](https://github.com/MetaMask/snaps-skunkworks/tree/main/packages/snap-examples) maintained by MetaMask.
+1. Extend the dapp-facing MetaMask JSON-RPC API in arbitrary ways.
+2. Integrate with and extend the functionality of MetaMask using the [snaps RPC methods and permissions](./snaps-rpc-api.html).
 
-Finally, you may wish to check out the first functional key management snap ever made, [the Filecoin Snap, or `filsnap`](https://github.com/Chainsafe/filsnap/).
+In this section, we'll go into detail about how to actually develop a snap and overcome common issues encountered during development.
 
-## Gotchas
+### The Snap Lifecycle
 
-### Accessing the Internet in a Snap
+Before beginning the development of your snap, it's important to understand the snap lifecycle.
+Just like [service workers](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) or AWS lambda functions, snaps are designed to wake up in response to messages / events, and shut down when they are idle.
+We say that snaps have an "ephemeral" lifecycle; here one moment, gone the next.
+In addition, if MetaMask detects that a snap becomes unresponsive, it will be shut down.
+This does not mean that you can't create long-running snaps, but it does mean that your snaps must handle being shut down, especially when they are not within the JSON-RPC request / response cycle.
+
+A snap is considered "unresponsive" if:
+
+1. It has not received a JSON-RPC request for 30 seconds.
```

Comment:
This only takes effect after the initial run of the snap (initiation), if the initiation never finishes the snap will run without time limits. See https://github.com/MetaMask/snaps-skunkworks/blob/ritave/onslaught/packages/snap-examples/examples/onslaught/src/index.js